### ISSUE TITLE
 some restrictions only allow certain ports to be forwarded to the server

### ISSUE THREAD

**Issue by huddhudd** [ntop/n2n#412]
I am very excited to see that N2n has been updating.
I deployed the n2n server a month ago. More than 5000 connections.
There is no traffic, but only the broadcast package will take up the server bandwidth of 1000M.

So I want to know if there is a way to deal with a problem
For example, some restrictions only allow certain ports to be forwarded to the server

**Comment by Logan007** [comment]
If it is broadcast packets that cause the high bandwidth usage, you can try to **not** give `-E` option at the edges.

By the way, 1,000 Mbit/s fully consumed by broadcast packets sounds like a whole lot of broadcasting apps – even for 5,000 nodes.

As for filtering traffic by ports, I have not seen such an option in n2n.

**Comment by huddhudd** [comment]
I want to build a P2P network. Open security for all ports is also an issue

**Comment by huddhudd** [comment]
Maybe it's my application scenario

N2n should not be used in this way

**Comment by Logan007** [comment]
It is just like you would use a VPN, n2n provides a full virtual network port. If you want to restrict it to the use by one application only, it has to happen on the client side as with all your other "regular" network ports. Perhaps `iptables` or local firewall configuration will help you to prevent applications from using certain network interfaces.

If you are still coding your P2P network, you could also think of integrating n2n into your application giving you more control over the data that is being sent. Please see `example_edge_embed.c` and others in the `src/` directory. By the help of the call-back hooks you will be able to discard those packets that you do not want transmitted via your channel.

By the way, you can tunnel certain IP addresses / networks only using the `-n` parameter – but that probably will not keep the broadcast down. Did it help to disable it not using `-E`?

Finally, we expect the communication architecture to see some upcoming changes. A distributed peer-to-peer basis is being discussed. That would relieve the central supernode.

**Comment by huddhudd** [comment]
thank you for your patience

I prefer to add a restriction on the inflow and outflow ports, I believe there are many people like me

**Comment by Logan007** [comment]
> I prefer to add a restriction on the inflow and outflow **ports**, …

Let me add as a post script that this would require n2n to evaluate and interfere with payload characteristics which is not within its scope… With a view to payload, n2n does not care about protocol, ports and stuff as long as the payload can be put into an ethernet frame. In this respect, it is neutral.

**Comment by huddhudd** [comment]
You can have this.

**Comment by Logan007** [comment]
Honestly, it would duplicate firewall functionalities.

But if you still want to customize your fork towards this direction anyway, you might consider using `edge_set_callbacks()` to register a `cb.packet_from_tap()` callback function helpful – please see `src/edge_utils.c`. Maybe `edge_init()` would be a good place to register the callback function.

Inside the custom callback function, your code can inspect the ethernet packet data `eth_pkt` along your needs (protocol, destination port, magic content bytes, any other criterion) and decide if to let pass or to drop the packet just by returning `N2N_DROP`.

Just keep in mind that `eth_pkt` is ethernet frame packet format (data link layer?).

**Comment by huddhudd** [comment]
I'm very sorry. I'm not very good at compiling programs.

I believe most people are the same. If you want more people to use n2n, I think we can add this function.

**Comment by Logan007** [comment]
> I think we can add this function.

I see.

So, to give it a start please find this real quick [patch](https://github.com/ntop/n2n/files/5220536/callback.txt). You can apply it to the source using `patch -p0 < callback.txt`.

It is a very simple filter for packets aiming at port 30,584. To tell the whole story, it only allows packets containing hex 0x7777 (=30,584) at some point. It is extremely hacky and probably only works with linux as `memmem()` is limited to some operating systems only. But it shows the basic idea of how the callbacks work and might serve as a good starting point to lift off from.

I have to make excuses for the poor quality of the code in the patch. Is there a network expert who could enhance it with some ethernet frame parsing, IP dissection, or so?

Just out of interest, did removing `-E` from the edges help to keep the idle traffic low?

By the way, I found that #18 might be related. We could solve it in the course of this issue too!

_P. S._ Not sure about that but we might need to find a way that lets ARP and ICMP packets pass untouched.

**Comment by Logan007** [comment]
If we check for the [ARP packet format](https://en.wikipedia.org/wiki/Address_Resolution_Protocol) (essential for building the network), we will find that the hex pattern `08 06 00 01 08 00 06 04` will recur (as long as embedded in ethernet frames) starting at position 12 of the outer ethernet frame. We could add it to the checking procedure and [update the patch](https://github.com/ntop/n2n/files/5225170/callback.txt) accordingly.

This will allow the nodes to find each other on system level. However, ICMP (used for ping) is embedded in IPv4 packet which again is embedded in ethernet. So, this is more complicated and `ping <ip address of other node>` does not work out of the box. But, if we got the hex pattern `77 77` anywhere inside the ping packet (see above), it would work. Fortunately, ping allows for custom hex patterns: `ping -p 7777 <ip address of other node>` magically works!

Very hacky, I admit, and I guess the code quality is not sufficient to have this go into the _dev_ branch. But does it help to solve the issue?

**Comment by Logan007** [comment]
_P. S._ Finally, if we wanted to pursue a firewall based solution instead (which in my opinion still is the proper way to deal with it), we could follow [this example](https://serverfault.com/questions/933505/iptables-block-outgoing-on-port-allow-selected-ips) or take advantage of a more [general description](https://www.cyberciti.biz/faq/iptables-block-port/):

The command
```
sudo iptables -A OUTPUT -o <n2n-device name> -p tcp ! --dport <allowed destination port> -j DROP
```
blocks (almost) all our outgoing traffic on the named _n2n-device_, e.g. `n2n0` with the exception that it lets pass the traffic headed towards the _allowed destination port_, e.g. `30583`. Note that this command covers tcp only but it could easily be extended to udp, too. The `iptables` command can be added to a `ExecStartPost=` line (there can be several) of the `.service`-file (without `sudo` then) in a similar manner as [suggested](https://github.com/ntop/n2n/issues/352#issuecomment-669196027) by @tlsalex .

By the way, when playing around with `iptables`, keep the _reset_ command handy: `sudo iptables -F OUTPUT`

Please let us know which solution works best.

**Comment by Logan007** [comment]
How did you solve it? Any news?

**Comment by huddhudd** [comment]
There is no way to deal with it. Need to change the source code implementation. Or with the firewall that comes with the system

**Comment by joshuafc** [comment]
@Logan007 
I'm interested to add a filter support allow/block list using src ip:port, dest ip:port.

Is there any reference material about how to unpack these infomation from parameter `uint8_t *payload` at packet_from_peer/packet_from_tap.

Or is there some library to process 'ethernet frame packet format' as you mentioned, so I can get then into ICMP,UDP,TCP packets with  src ip:port, dest ip:port infomation.

**Comment by Logan007** [comment]
> Is there any reference material about how to unpack these infomation from parameter uint8_t *payload at packet_from_peer/packet_from_tap.

Those are [Ethernet frames](https://en.wikipedia.org/wiki/Ethernet_frame#Frame_%E2%80%93_data_link_layer) as seen on the data link layer. You can verify if you put `hexdump (payload, payload_size)` at some point in the code. IP with its own [headers](https://gohalo.me/reference/linux/network/IP_TCP_UDP_Header.pdf) (v4) for UDP or TCP, respectively, is encapsulated within. v6 has different offsets.

You would need to filter in several steps: First, only process IP packets (filter along [Ethertype](https://en.wikipedia.org/wiki/EtherType), watch out for [VLAN tags and double tags](https://en.wikipedia.org/wiki/IEEE_802.1Q#Double_tagging) which can change the packet offset for IP addresses and ports!) and let all others pass. Then, based on that, you will find the offsets pointing to data in the packets (addresses, ports) on which you can apply your filter rules on.

Easiest way to drop (or accept if non-dropped) packets is by using the callbacks as drafted [above](https://github.com/ntop/n2n/issues/412#issuecomment-692722675) in that patch.

Talking about the rules, you would need to find a syntax for command line and internal format to store the rules.

> I'm interested to add a filter support allow/block list using src ip:port, dest ip:port.

:+1:

For the sake of clarity, you might want to put your actual filtering callbacks into a seperate `.c` source file.

> Or is there some library to process 'ethernet frame packet format' as you mentioned, so I can get then into ICMP,UDP,TCP packets with src ip:port, dest ip:port infomation.

There surely are, I get tons of results on google while real quick searching for `packet processing library ip`. I just am not sure if we really want to load n2n with another external library for that purpose (which one still might see as firewall-duplication).

I hope these hints are somewhat helpful. Let us know!

**Comment by joshuafc** [comment]
@Logan007 Thank you for your guidance, now I understand the concept of Ethernet frames. And I had wrote some code process Ethernet frames successfully, after referring to some codes from a  sniffer project.

It's still a little difficult for me to add the feature to n2n. I mainly use C + + and Python to develop programs. As I see, n2n is a pure C language project, I will try wrote structs and functions for the internal format to store the rules.  To be honest, I'm not familiar with it. I used to use  array, list, map containers.

And, may be it is better to store the rules in a file, and specify it's path at command line when run edge. For the case there is too much rules to type at terminal. Is it acceptable to introduce a lightweight single file parser for ( json/yaml )?

**Comment by Logan007** [comment]
For internal rule representation, I would use a HASH_LIST of _rules_. n2n uses utHash all around for lists, for example for communities (as list of nodes). They are relatively easy to use. 

To represent a _rule_, you might need the following fields in a struct:
```
n2n_sock_t source
n2n_sock_t destination
uint16_t protocol
uint8_t actionOnMatch
uint8_t actionOnNoMatch
```

The `n2n_sock_t` already has fields for IP addresses (v4 and v6) and port (defined in `n2n_wire.h:102`). 

> For the case there is too much rules to type at terminal.

That is exactly what the `.conf` file is for – each line a parameter or rule. I would not add an extra file.

**Comment by joshuafc** [comment]
#489 Almost done! I would like to receive any suggestions to improve this feature.

**Comment by huddhudd** [comment]
I want to limit the bandwidth of each connection when forwarding data

**Comment by huddhudd** [comment]
For IP Address

**Comment by huddhudd** [comment]
For IP Address

**Comment by Logan007** [comment]
The original request has been fulfilled by @joshuafc . @huddhudd, if you still desire the bandwidth limit, please open a new issue.

**Comment by duhd1993** [comment]
@joshuafc Sorry to bother you. I'm a little confused about the setting. I have two edge nodes(192.168.100.1 or 2). I pass -R arguments on 192.168.100.1. My goal is to access .2 on .1 while restrict access to .1 from .2.

`-R 0.0.0.0/0,0.0.0.0/0,TCP-,UDP-,ICMP- -R 192.168.100.1/32,192.168.100.2/32,ICMP+` I thought this would allow me to ping 192.168.100.2 but it doesn't work. changing to /24 works.

`-R 0.0.0.0/0,192.168.100.1/32,TCP-,UDP-,ICMP-` I though this could allow outgoing from 192.168.100.1 while blocking incoming connection. But all connections are blocked.
 
`-R 0.0.0.0/0,0.0.0.0/0,TCP-,UDP-,ICMP- -R 192.168.100.0/30,TCP+` allows using tcp without issues. But `-R 0.0.0.0/0,0.0.0.0/0,TCP-,UDP-,ICMP- -R 192.168.100.0/30:3389,TCP+` when restricting port, it does not connect through.

**Comment by joshuafc** [comment]
@lonelykid rule for single ip, you can ignore '/32', I only tested this way. 

`-R 0.0.0.0/0,0.0.0.0/0,TCP-,UDP-,ICMP- -R 192.168.100.1/32,192.168.100.2/32,ICMP+` ICMP reply from .2 to .1 blocked.

`-R 0.0.0.0/0,192.168.100.1/32,TCP-,UDP-,ICMP-` blocking all packet from any address to `192.168.100.1`. So outgoing from 192.168.100.1 allowed, but reply at the same connection blocked.

.2:12345 --> .1:3389 and  .1:3389 --> .2:12345 all allowed with `-R 192.168.100.0/30,TCP+`.
but .2:12345 --> .1:3389 will blocked by `-R 192.168.100.0/30:3389`


The block rule not only for Request, But also work for Reply.

**Comment by joshuafc** [comment]
@lonelykid Maybe you can try like this:
`-R 0.0.0.0/0,192.168.100.1,TCP-,UDP-,ICMP- -R 0.0.0.0/0,192.168.100.1:3389,TCP+`
then except  packets to 192.168.100.1:3389, packet to 192.168.100.1 whill blocked. And all packets from 192.168.100.1 to and address are allowed.

**Comment by Logan007** [comment]
@huddhudd, with a view to #671, I would be interested if the filter rules help you to operate the 5,000 edges network.

Also, the federation feature might help to share the broadcast burden because edges get distributed (somewhat) equally to the available supernodes. That way, the supernodes also share the burden of forwarding broadcasts.

**Comment by KusakabeSi** [comment]
Consider this?
![image](https://user-images.githubusercontent.com/73118488/111958257-f8421d80-8b27-11eb-8eeb-420c0d21fbc0.png)
split your network into multiple smaller network.

**Comment by duhd1993** [comment]
@joshuafc Thank you. That's clear to me. In summary, it's not able to distinguish an incoming ping/tcp packet and a reply of an outgoing ping/tcp packet. Otherwise, a reply should always be allowed IMO. Is this possible?
```
-R 0.0.0.0/0,192.168.100.1,TCP-,UDP-,ICMP- 
-R 0.0.0.0/0,192.168.100.1:3389,TCP+
```
This does not work because the port used by .1 is not 3389; it's not even fixed in general. I think I can do `0.0.0.0/24,192.168.100.1:[10000,65535],TCP+` the large number ports are used for outgoing tcp connection. all the sensitive ports are blocked.
